test/functional/p2p_ibd_stalling.py

```diff
@@ -0,0 +1,145 @@
+            self.getdata_requests.append(inv.hash)
+            if (inv.type & MSG_TYPE_MASK) == MSG_BLOCK:
+                if(not self.is_stalling or inv.hash != self.stall_block):
+                    self.send_message(msg_block(self.block_store[inv.hash]))
+                    self.blocks_sent += 1
+
+    def on_getheaders(self, message):
+        pass
+
+
+class P2PIBDStallingTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 1
+
+    def run_test(self):
+        NUM_BLOCKS = 1025
+        node = self.nodes[0]
+        tip = int(node.getbestblockhash(), 16)
+        blocks = []
+        height = 1
+        block_time = node.getblock(node.getbestblockhash())['time'] + 1
+        self.log.info("Prepare blocks without sending to the node")
+        block_dict = {}
+        for _ in range(NUM_BLOCKS):
+            blocks.append(create_block(tip, create_coinbase(height), block_time))
+            blocks[-1].solve()
+            tip = blocks[-1].sha256
+            block_time += 1
+            height += 1
+            block_dict[blocks[-1].sha256] = blocks[-1]
+
+        stall_block = blocks[0].sha256
+
+        headers_message = msg_headers()
+        headers_message.headers = [CBlockHeader(b) for b in blocks[:NUM_BLOCKS-1]]
+        peers = []
+
+        self.log.info("Check that stalling does not start the stalling logic yet if there is a 1024 block window.")
+        with self.nodes[0].assert_debug_log([], unexpected_msgs=['Stall started']):
+            for id in range(8):
+                peers.append(node.add_outbound_p2p_connection(P2PStaller(stall_block), p2p_idx=id, connection_type="outbound-full-relay"))
+                peers[-1].block_store = block_dict
+                peers[-1].send_message(headers_message)
+            self.wait_until(lambda: self.total_blocks_sent(peers) == NUM_BLOCKS - 2)
+        time.sleep(0.5)  # Wait until all blocks have arrived at the node
+
+        self.log.info("Check that increasing the window beyond 1024 blocks triggers stalling logic")
+        headers_message.headers = [CBlockHeader(b) for b in blocks]
+        with self.nodes[0].assert_debug_log(expected_msgs=['Stall started peer=0']):
+            for p in peers:
+                p.send_message(headers_message)
+
+        self.log.info("Check that the stalling peer is disconnected after 2 seconds")
+        self.mocktime = int(time.time()) + 3
+        node.setmocktime(self.mocktime)
+        peers[0].wait_for_disconnect()
+        assert_equal(len(self.nodes[0].getpeerinfo()), 7)
+        self.wait_until(lambda: self.is_block_requested(peers, stall_block))
+        # Make sure that SendMessages() is invoked, which assigns the missing block
+        # to another peer and start the stalling logic for them
+        self.all_sync_send_with_ping(peers)
+
+        self.log.info("Check that the stalling timeout gets doubled to 4 seconds for the next staller")
```

Comment:
This test actually almost passes for the previous stalling logic as well, it only fails on `Check that stalling timeout gets reduced back to 2 seconds after connecting several blocks` (because of the assertion on the debug log).

Maybe add a condition to check that after 3 seconds the stalling peer is not disconnected? (same for the next doubling to 8 seconds)

Reply:
Done with the latest push.